solo-chains/runtime/starlight/src/lib.rs

```diff
@@ -948,22 +947,102 @@ impl parachains_paras_inherent::Config for Runtime {
 impl parachains_scheduler::Config for Runtime {
     // If you change this, make sure the `Assignment` type of the new provider is binary compatible,
     // otherwise provide a migration.
-    type AssignmentProvider = CoretimeAssignmentProvider;
+    type AssignmentProvider = CollatorAssignmentProvider;
 }
 
-parameter_types! {
-    pub const BrokerId: u32 = BROKER_ID;
-    pub MaxXcmTransactWeight: Weight = Weight::from_parts(200_000_000, 20_000);
-}
+pub struct CollatorAssignmentProvider;
+impl parachains_scheduler::common::AssignmentProvider<BlockNumberFor<Runtime>>
+    for CollatorAssignmentProvider
+{
+    fn pop_assignment_for_core(core_idx: CoreIndex) -> Option<Assignment> {
+        let assigned_collators = TanssiCollatorAssignment::collator_container_chain();
+        let assigned_paras: Vec<ParaId> = assigned_collators
+            .container_chains
+            .iter()
+            .filter_map(|(&para_id, _)| {
+                if Paras::is_parachain(para_id) {
+                    Some(para_id)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        log::info!("pop assigned collators {:?}", assigned_paras);
+        log::info!("looking for core idx {:?}", core_idx);
+
+        if let Some(para_id) = assigned_paras.get(core_idx.0 as usize) {
+            log::info!("outputing assignment for  {:?}", para_id);
+
+            Some(Assignment::Bulk(*para_id))
+        } else {
+            // We dont want to assign affinity to a parathread that has not collators assigned
+            // Even if we did they would need their own collators to produce blocks, but for now
+            // I prefer to forbid.
+            // In this case the parathread would have bought the core for nothing
+            let assignment =
+                parachains_assigner_on_demand::Pallet::<Runtime>::pop_assignment_for_core(
+                    core_idx,
+                )?;
+            if assigned_collators
+                .container_chains
+                .contains_key(&assignment.para_id())
```

Comment:
I don't understand this check. If `assigned_paras.get` returns `None`, shouldn't we just return `Assignment::Pool`?

Reply:
the reason for this check is because people can buy (as of today) a core for a parathread that does not have collators assigned. Just for sanity purposes I would rather have assignments returned for just parathreads that have collators.

It is not going to change anything in the end, but at least is should avoid trying to contact non-existing collators for parathreads

Reply:
Maybe this does not play well with the fact that you can have your own collators. But I doubt you would want to use starlight in that case?

Reply:
It's weird that we do this after calling `pop_assignment_for_core(core_idx)`. Because if a parathread buys core number 0, it will be ignored forever? (as long as we have parachains, because assigned_paras.get(0) will return Some and never enter this branch)

So it feels to me that this way of handling the assignment is a bit too hacky. Does `parachains_assigner_on_demand` know that the first n cores are used by parachains?

Reply:
parathreads cannot buy core 0. in fact they do not know what core they buy. their assignment simply go to  a queue.

this queue is later handled by this code. this is for every non-parachain occupied core, we pop one of the assignments from that queue and they occupy the selected core

Reply:
Here the code: https://github.com/paritytech/polkadot-sdk/blob/6a5b6e03bfc8d0c6f5f05f3180313902c15aee84/polkadot/runtime/parachains/src/on_demand/mod.rs#L284

Reply:
Ok so basically this is not an issue because any core affinity gets reset on session changes, as seen in test `test_on_demand_core_affinity_bound_to_core_gets_expired_at_session_boundaries`